From: Martin Schwidefsky <schwidefsky@de.ibm.com>
Subject: s390: fix test_fp_ctl inline assembly contraints
Patch-mainline: v4.7-rc6
Git-commit: bcf4dd5f9ee096bd1510f838dd4750c35df4e38b
References: bnc#988217, LTC#143138

Description:  kernel: signal return with invalid floating-point control
Symptom:      A specification exception occurs in the load_fpu_regs
              kernel function if a user space process returns from a
              signal with an invalid floating-point control register
              in the signal frame stored on the user stack. With
              panic_on_oops=1 the system panics.
Problem:      The signal return code uses the test_fp_ctl function to test
              if the floating-point control value retrieved from the user
              user stack is valid. The inline assembly in test_fp_ctl uses
              an incorrect constraint. If the kernel has been compiled
              with gcc 4.8.5 this causes test_fp_ctl to return true even
              for invalid floating-point control values.
Solution:     Correct the inline assembly contraint for orig_fpc in the
              test_fp_ctl function.
Reproduction: While the user program is inside a signal handler write an
              invalid floating-point control value to the signal frame on
              the user stack and then return from the signal handler.

Upstream-Description:

              s390: fix test_fp_ctl inline assembly contraints

              The test_fp_ctl function is used to test if a given value is a valid
              floating-point control. The inline assembly in test_fp_ctl uses an
              incorrect constraint for the 'orig_fpc' variable. If the compiler
              chooses the same register for 'fpc' and 'orig_fpc' the test_fp_ctl()
              function always returns true. This allows user space to trigger
              kernel oopses with invalid floating-point control values on the
              signal stack.

              This problem has been introduced with git commit 4725c86055f5bbdcdf
              "s390: fix save and restore of the floating-point-control register"

              Cc: stable@vger.kernel.org # v3.13+
              Reviewed-by: Heiko Carstens <heiko.carstens@de.ibm.com>
              Signed-off-by: Martin Schwidefsky <schwidefsky@de.ibm.com>


Signed-off-by: Martin Schwidefsky <schwidefsky@de.ibm.com>
Acked-by: John Jolly <jjolly@suse.de>
---
 arch/s390/include/asm/fpu/api.h |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

--- a/arch/s390/include/asm/fpu/api.h
+++ b/arch/s390/include/asm/fpu/api.h
@@ -22,7 +22,7 @@ static inline int test_fp_ctl(u32 fpc)
 		"	la	%0,0\n"
 		"1:\n"
 		EX_TABLE(0b,1b)
-		: "=d" (rc), "=d" (orig_fpc)
+		: "=d" (rc), "=&d" (orig_fpc)
 		: "d" (fpc), "0" (-EINVAL));
 	return rc;
 }
